src/job.rs

```diff
@@ -2,32 +2,78 @@ use crate::glue;
 use anyhow::{Context, Result};
 use itertools::Itertools;
 use roc_std::RocStr;
-use std::collections::{HashMap, HashSet};
+use std::collections::HashSet;
 use std::fmt::{self, Display};
 use std::hash::{Hash, Hasher};
-use std::path::{Component, PathBuf};
+use std::marker::PhantomData;
+use std::path::{Component, Path, PathBuf};
 use std::process::Command;
 use xxhash_rust::xxh3::Xxh3;
 
+/// Conversion from a base key to a final one.
+pub struct KeyBuilder(Xxh3);
+
+impl KeyBuilder {
+    fn new() -> Self {
+        Self(Xxh3::new())
+    }
+
+    pub fn based_on(id: &Key<Base>) -> Self {
+        let mut builder = Self::new();
+        id.hash(&mut builder.0);
+
+        builder
+    }
+
+    pub fn add_file(&mut self, path: &Path, content_hash: &str) {
+        path.hash(&mut self.0);
+        content_hash.hash(&mut self.0);
```

Comment:
I thought we were avoiding hashing file content and just hashing the metadata, but I am probably just confused from comments elsewhere to use here.